proposals/webidl-bindings/idl.md

```diff
@@ -0,0 +1,739 @@
+data value may refer to entities that have interfaces around them.
+
+### Nominal vs Structural Types {#nominal-types}
+
+A nominal type is one whose semantics depends on a denotation: on the entities
+being denoted by the type. For example, a Person type is typically understood as
+meaning or denoting people -- not by any internal structure associated with
+Person values.
+
+On the other hand, a structural type is one whose semantics arises from
+composition: from the components of its internal representation. For example, a
+tuple pair type denotes pairs of values; it is not expected that tuple pairs
+would be used to model ‘real world’ entities.
+
+Structural types are important for what one can do with them, nominal types are
+important for what they mean.
+
+In this schema, nominal types are modeled in terms of type quantification: a
+nominal type is one that is introduced via a quantifier in a quantified
+type. Operationally, the difference between a nominal type and a structural type
+is that the former is fundamentally opaque whereas a structural type is
+transparent: you can see the pieces of a structural type.
+
+### Algebraic Types {#algebraic-types}
+
+An algebraic type is one whose values are defined by composition of two
+fundamental operators: tupling and union. Algebraic types are important as a way
+of expressing alternate forms of data; two examples being enumerations (such as
+boolean) and alternate representations of a point (in terms of cartesian or
+polar coordinates).
+
+### Type Quantification {#type-quantification}
+
+A quantified type is a way of denoting an arbitrary number of types -- in the
+case of universal quantifiers an infinite number and in the case of existential
+quantifiers at least one. In addition to this enumeration, quantified types also
+allow us to clarify scope; in particular nested scope.
+
+In addition to quantifying types, it is also possible for a type to be
+generic. A generic type is like a function on types: it takes a (typically one)
+type as argument and returns a new type.
+
+Generic types have two roles in this schema: they underlie the built-in concepts
+of Lists and Maps; and they can be used to model certain ‘user-defined’ types --
+such as a Permission type which might need to have the type of permission
+‘plugged in’.
+
+### Type Equality {#type-equality}
+
+Compatibility between types is defined to be based on equality. In particular,
+subtyping is not supported by this schema.
+
+The reason for this is that subtypes introduce significant complexity for
+validation; furthermore, when combined with binding expressions whose task it is
+to raise or lower values to meet the type signature of an operation, subtyping
+is unnecessary.
+
+### Collections {#collections}
+
+The two standard collection types in this schema are the sequence and the
+map. This allows complex data -- such as a sequence of points each of which is
```

Comment:
Are maps needed as a primitive concept? Is it more than just a sequence of key/value pairs? The only way I could see the notion of primitive map being a meaningful addition was if the IDL ensures that it's *actually* a map, i.e., keys are unique. But that would be expensive both in terms of semantics (you'd need to specify value equality) as well as runtime (who enforces this invariant?).